packages/grafana-ui/src/components/DateTimePickers/DateTimePicker/DateTimePicker.tsx

```diff
@@ -277,68 +284,50 @@ const DateTimeCalendar = React.forwardRef<HTMLDivElement, DateTimeCalendarProps>
-    const getFormattedDate = useGetFormattedDate(showSeconds);
-
-    // To simply the handing of `DateTime` and `Date` objects, we will use store them into separate states
-    // and as strings. This way, we can easily convert them to `DateTime` objects when needed (see `value`).
-    const [timeValue, setTimeValue] = useState<string>(getFormattedDate(date && date.isValid() ? date : dateTime()));
-    const [dateValue, setDateValue] = useState<string>(timeValue);
-
-    useEffect(() => {
-      const newValue = getFormattedDate(date);
-      setTimeValue(newValue);
-    }, [getFormattedDate, date]);
-
-    useEffect(() => {
-      setDateValue(timeValue);
-    }, [timeValue]);
-
-    useEffect(() => {}, []);
-
-    // Create date time object from the date and time values
-    // This is both used by `TimeOfDayPicker.value` and as param for the `onChange` callback
-    const value = useMemo(() => {
-      const [datePart] = dateValue.split(' ');
-      const [, timePart] = timeValue.split(' ');
+    const [internalDate, setInternalDate] = useState<Date>(() => {
+      if (date && date.isValid()) {
+        return date.toDate();
+      }
 
-      // Combine the date and time parts and apply user timezone to get the correct `DateTime` object
-      return dateTimeForTimeZone(getTimeZone(), `${datePart} ${timePart}`);
-    }, [dateValue, timeValue]);
+      return new Date();
+    });
 
-    const handleOnChangeDate = useCallback<NonNullable<React.ComponentProps<typeof Calendar>['onChange']>>(
-      (newValue) => {
-        if (Array.isArray(newValue)) {
-          return;
-        }
+    const onChangeDate = useCallback<NonNullable<React.ComponentProps<typeof Calendar>['onChange']>>((date) => {
+      if (date && !Array.isArray(date)) {
+        setInternalDate((prevState) => {
+          // If we don't use time from prevState
+          // the time will be reset to 00:00:00
+          date.setHours(prevState.getHours());
+          date.setMinutes(prevState.getMinutes());
+          date.setSeconds(prevState.getSeconds());
 
-        // Calendar is using the browser timezone, so we cannot apply user timezone when formatting
-        setDateValue(dateTimeFormat(newValue, { timeZone: 'browser' }));
-      },
-      []
-    );
-
-    const handleOnChangeTime = useCallback(
-      (newValue: DateTime) => {
-        setTimeValue(getFormattedDate(newValue));
-      },
-      [getFormattedDate]
-    );
+          return date;
+        });
+      }
+    }, []);
 
-    const handleOnApply = useCallback(() => {
-      onChange(value);
-    }, [onChange, value]);
+    const onChangeTime = useCallback((date: DateTime) => {
+      setInternalDate((prevState) => {
+        const newDate = new Date(prevState);
+        const newTime = date.toDate();
+        newDate.setHours(newTime.getHours());
+        newDate.setMinutes(newTime.getMinutes());
+        newDate.setSeconds(newTime.getSeconds());
+        return newDate;
+      });
+    }, []);
```

Comment:
`DateTime` is only a small abstraction around `Date`. they're basically interchangeable. i don't think we're mutating, we're just setting a new state 🤷 

i see a few benefits:
- don't have to keep 2 state variables in sync (_this_ will cause bugs)
- keeps the delta between this and main smaller 🙏 
- as you said, we have to do this syncing anyway, why not do it in state?

Reply:
`DateTime` is `Moment as DateTime`, right? Don't think they are interchangeable. 🤔 
I never meant that the state was mutated (which it isn't). I was referring to the use of `setHours|setMinutes|setSeconds`. Was a bad choice of words in the context of react state 😅 Sorry for the confusion.  

> don't have to keep 2 state variables in sync (this will cause bugs)

They don't need to be in sync, as the date shouldn't affect the time and vice versa (SoC). By treating it as one point in time, when they in reality may (at least should) be two points in time, is what I believe is the brittle part.
(warning, poor comparison) I'm sure we wouldn't try to use one state for two text inputs.

> keeps the delta between this and main smaller 🙏

👍🏻 

> as you said, we have to do this syncing anyway, why not do it in state?

Mainly because one component uses `Date` and client timezone, whilst the other component uses `Moment/DateTime` (and may be impacted by a timezone that differs from the client timezone).

Reply:
@w1kman yeah i'm beginning to agree with you, just needed time to get there 😂 

i'm about to push changes, also found a util function in another time picker that can help 👍